File: algorithms/go/linkedlist/linked_list.go
package main

////////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2019, Gianluca Fiore
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
////////////////////////////////////////////////////////////////////////////////

import (
	"fmt"
)

type Node struct {
	data      interface{}
	next_node *Node
}

type LinkedList struct {
	head *Node
	tail *Node
	size int
}

func newNode(val interface{}) *Node {
	return &Node{val, nil}
}

func (n *Node) getData() interface{} {
	return n.data
}

func (n *Node) setData(val interface{}) {
	n.data = val
}

func newLinkedList() *LinkedList {
	return &LinkedList{nil, nil, 0}
}

func (ll *LinkedList) getSize() int {
	return ll.size
}

func (ll *LinkedList) printList() {
	if ll.head == nil {
		fmt.Println("Empty List")
		return
	}

	fmt.Printf("List of %d\n", ll.size)
	curr := ll.head
	for {
		fmt.Println(curr.getData())
		curr = curr.next_node
		if curr == nil {
			break
		}
	}
}

func (ll *LinkedList) insert(n *Node) {
	if ll.head == nil {
		ll.head = n
	} else {
		ll.tail.next_node = n
	}

	ll.tail = n
	ll.size += 1
}

func (ll *LinkedList) traverse() []interface{} {
	var y []interface{}
	n := ll.head
	for {
		y = append(y, n.getData())
		n = n.next_node
		if n == nil {
			break
		}
	}

	return y
}

func (ll *LinkedList) reverseTraverse() []interface{} {
	var y []interface{}
	var curr, prev *Node
	if ll.tail != nil {
		curr = ll.tail
		for curr != ll.head {
			prev = ll.head
			for prev.next_node != curr {
				prev = prev.next_node
			}
			y = append(y, curr.data)
			curr = prev
		}
		y = append(y, curr.data)
		return y
	} else {
		y = append(y, ll.head.data)
		return y
	}
}

func (ll *LinkedList) contains(v interface{}) bool {
	curr := ll.head
	for curr != nil {
		if curr.getData() == v {
			fmt.Println("Value is present")
			return true
		}
		curr = curr.next_node
	}
	fmt.Println("Value is absent")
	return false
}

func (ll *LinkedList) remove(n *Node) bool {
	var curr *Node
	if ll.head == nil {
		// Empty list
		return false
	}
	curr = ll.head
	if curr.data == n.data {
		if ll.head == ll.tail {
			// there's only the head in the list, remove it
			ll.head = nil
			ll.tail = nil
		} else {
			// node is the head, change head then to the next node in list
			ll.head = ll.head.next_node
		}
		ll.size -= 1
		return true
	}
	for curr.next_node != nil && curr.next_node.data != n.data {
		// iterate on the list
		curr = curr.next_node
	}

	if curr.next_node != nil {
		if curr.next_node == ll.tail {
			// node is the tail
			ll.tail = curr
		}
		// node is between head and tail, so skip it to remove
		curr.next_node = curr.next_node.next_node
		ll.size -= 1
		return true
	}
	// if all the tests fail, node is not in the list
	return false
}


func main() {
	ll := newLinkedList()
	fmt.Println(ll.size)
	n1 := newNode(44)
	n2 := newNode("ave")
	n3 := newNode(8.8)
	ll.insert(n1)
	ll.insert(n2)
	ll.insert(n3)
	fmt.Println("Original list")
	ll.printList()
	ll.remove(n3)
	fmt.Println("List removed")
	ll.printList()
}

File: algorithms/python/linkedlist/linked_list.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
###############################################################################
#
# Copyright (c) 2019, Gianluca Fiore
#
###############################################################################

__author__ = "Gianluca Fiore"

# Big Oh on Linked Lists:
# Insertion = O(1) if the node to be inserted is the head or tail, otherwise it's O(n)
# Deletion = O(n)
# Searching = O(n)

class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node

    # Return the current Node's value
    def getData(self):
        return self.data

    # Set a new value for Node
    def setData(self, val):
        self.data = val

class LinkedList(object):
    def __init__(self, head=None, tail=None):
        self.head = head
        self.tail = tail
        self.size = 0

    # Print all values of nodes in the list
    def printList(self):
        if self.head == None:
            print("Empty List")
            return

        print("List of %i nodes" % self.size)
        curr = self.head
        while curr:
            print(curr.getData())
            curr = curr.next_node

    # Get the current size of the list
    def getSize(self):
        return self.size

    # Insert a new Node in the list
    def insert(self, node):
        if self.head == None:
            self.head = node
        else:
            self.tail.next_node = node

        self.tail = node
        self.size += 1

    # Traverse the list and yield every data in it
    def traverse(self):
        n = self.head
        while n != None:
            yield n.data
            n = n.next_node

    # Traverse the list from the tail
    def reverseTraverse(self):
        if self.tail != None:
            curr = self.tail
            while curr != self.head:
                prev = self.head
                while prev.next_node != curr:
                    prev = prev.next_node
                yield curr.data
                curr = prev
            yield curr.data

    # Check if a value is present in the list or not
    def contains(self, value):
        n = self.head
        while n:
            if n.getData() == value:
                print("Value is present")
                return True
            n = n.next_node

        print("Value is absent")
        return False

    # Remove a node from the list
    def remove(self, node):
        if node is None or node.next_node is None:
            raise ValueError
        node.data = node.next_node.data
        node.next_node = node.next_node.next_node

        self.size -= 1

def main():
    # Test functionalities
    l = LinkedList()
    new_node = Node(13)
    second_node = Node("abc")
    third_node = Node(8.4)

    l.insert(new_node)
    l.insert(second_node)
    l.insert(third_node)
    l.printList()
    l.contains("abc")

if __name__ == '__main__':
    main()
